scripts/gb-model/compose_network.py

```diff
@@ -0,0 +1,311 @@
+            and settings.get("landfall_length") is not None
+        }
+        line_length_factor = config.get("lines", {}).get("length_factor", 1.0)
+
+        attach_wind_and_solar(
+            network,
+            costs,
+            ppl,
+            {k: str(v) for k, v in profile_paths.items()},
+            available_non_hydro,
+            extendable_carriers,
+            line_length_factor,
+            landfall_lengths,
+        )
+
+    if "hydro" not in renewable_carriers:
+        return
+
+    hydro_profile = profile_paths.get("profile_hydro")
+    if hydro_profile is None or not hydro_profile.exists():
+        logger.warning("Hydro profile not available; skipping hydro integration")
+        return
+
+    if hydro_capacities_path is None or not hydro_capacities_path.exists():
+        logger.warning(
+            "Hydro capacities file %s missing; skipping hydro integration",
+            hydro_capacities_path,
+        )
+        return
+
+    hydro_cfg = copy.deepcopy(config.get("renewable", {}).get("hydro", {}))
+    carriers = hydro_cfg.pop("carriers", [])
+
+    attach_hydro(
+        network,
+        costs,
+        ppl,
+        str(hydro_profile),
+        str(hydro_capacities_path),
+        carriers,
+        **hydro_cfg,
+    )
+
+
+def add_gb_components(
+    n: pypsa.Network,
+    context: CompositionContext,
+) -> pypsa.Network:
+    if context.countries:
+        keep = n.buses.country.isin(context.countries)
+        drop = n.buses.index[~keep]
+        if len(drop) > 0:
+            logger.info("Removing %d buses outside target countries", len(drop))
+            n.mremove("Bus", drop)
+
+    meta = n.meta.setdefault("gb_model", {})
+    if context.countries:
+        meta["countries"] = list(context.countries)
+
+    return n
+
+
+def add_pypsaeur_components(
+    n: pypsa.Network,
+    config: dict[str, Any],
+    context: CompositionContext,
+    costs,
+) -> pypsa.Network:
+    if costs is not None:
+        add_electricity_grid_connection(n, costs)
+        n.meta.setdefault("gb_model", {})["costs_path"] = str(context.costs_path)
+
+    sanitize_locations(n)
+    try:
+        sanitize_carriers(n, config)
+    except KeyError as exc:  # pragma: no cover - tolerate partial configs
+        logger.debug("Skipping carrier sanitisation due to missing config: %s", exc)
+    return n
+
+
+def finalise_composed_network(
+    n: pypsa.Network,
+    context: CompositionContext,
+) -> pypsa.Network:
+    n.determine_network_topology()
+    meta = n.meta.setdefault("gb_model", {})
+    meta["resources_root"] = str(context.resources_root)
+    meta["composed"] = True
+    n.consistency_check()
+    return n
+
+
+def compose_network(
+    network_path: str,
+    output_path: str,
+    config: dict[str, Any],
+    inputs: Mapping[str, str],
+) -> None:
+    network = pypsa.Network(network_path)
+    context = create_context(network_path, config)
+    if "tech_costs" in inputs:
```

Comment:
unnecessary? `tech_costs` _will_ be in inputs as there is only one rule calling this script